libraries/chain/hotstuff/hs_pseudo

```diff
@@ -0,0 +1,345 @@
+		//my last vote was on (t9, t10_1], I'm asked to vote on t10 : t9 < t10 && t9 < t10_1; //time_range_interference == true, correct
+		//my last vote was on (t9, t10_1], I'm asked to vote on t11 : t9 < t11 && t10 < t10_1; //time_range_interference == false, correct
+		//my last vote was on (t7, t9], I'm asked to vote on t10 : t7 < t10 && t9 < t9; //time_range_interference == false, correct
+		bool enough_for_strong_vote = false;
+		if (!time_range_interference || extends(p, fork_db.get_block_by_id(fsi.last_vote_block_ref)) enough_for_strong_vote = true;
+		fsi.is_last_vote_strong = enough_for_strong_vote;
+		fsi.last_vote_block_ref = p.block_id; //v_height
+		if (b1.timestamp > fork_db.get_block_by_id(fsi.locked_block_ref).timestamp) fsi.locked_block_ref = b1.block_id; //commit phase on b1
+		fsi.last_vote_range_lower_bound = new_vote_range_lower_bound;
+		fsi.last_vote_range_upper_bound = new_vote_range_upper_bound;
+		if (enough_for_strong_vote) return VoteDecision::StrongVote;
+		else return VoteDecision::WeakVote;
+	}
+	else  return VoteDecision::NoVote;
+}
+
+//handlers
+
+void on_signed_block_received(signed_block sb){
+	[...] //verify if block can be linked to our fork database, throw exception if unable to or if duplicate
+	block_handle previous = fork_db.get_block_by_id(sb.previous);
+	hs_proposal p = extract_proposal(sb, previous);
+	on_proposal_received(p, previous);
+}
+
+void on_proposal_received(hs_proposal p, block_handle& bc){
+
+	//relevant to all nodes
+	if (p.last_qc_block_height > bc.bhs.last_qc_block_height) {
+		block_handle found = fork_db.get_block_by_height(p.block_id, p.last_qc_block_height);
+		//verify qc is present and if the qc is valid with respect to the found block, throw exception otherwise
+	}
+
+	core new_core = bc.bhs.core.next(p.last_qc_block_height, p.is_last_qc_strong, ...); //f1
+	//todo : f2 
+	[...] // add to fork db + update block_header_state.core
+
+	bls_public_key[] my_finalizers = [...] //abstracted, must return the public keys of my finalizers that are also active in the current finalizer policy
+	//only relevant if I have at least one finalizer
+	if (my_finalizers.size()>0) {
+		for (auto f : my_finalizers){
+			finalizer_safety_information& fsi = get_finalizer_info(f);
+			vote_decision vd = decide_vote(p, fsi, bc); //changes fsi unless NoVote
+			if (vd == VoteDecision::StrongVote || vd == VoteDecision::WeakVote){
+				save_finalizer_info(f, fsi); //save finalizer info to prevent double-voting
+				hs_vote_message msg = [...] //create + broadcast vote message
+			}
+		}
+	}
+}
+
+//when a node receives a vote on a proposal
+void on_vote_received(hs_vote_message v){
+	//check for duplicate or invalid vote, return in either case
+	//abstracted [...]
+	block_handle& bc = fork_db.get_block_by_id(v.block_id);
+	am_i_leader = [...] //abstracted, must return true if I am the leader, false otherwise
+	if(!am_i_leader) return;
+	//only leader need to take action on votes
```

Comment:
Need to relay votes even if not a leader.

Reply:
I added a comment to mention this. Also added a similar comment on_proposal_received for consistency.